project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt

```diff
@@ -1,33 +1,110 @@
                 val vf = it.getVirtualFile(minimizationResult.context) ?: return@flatMap emptyList()
-                vf.children.map { file ->
-                    ProjectFileDDItem.create(minimizationResult.context, file)
-                }
+                readAction { vf.children }
+                    .map { file ->
+                        ProjectFileDDItem.create(minimizationResult.context, file)
+                    }
             }
             ensure(nextFiles.isNotEmpty())
+
+            reporter?.updateProgress(nextFiles)
             HDDLevel(minimizationResult.context.copy(currentLevel = nextFiles), nextFiles, propertyTester)
         }
+
+    /**
+     * ProgressReporter is a class responsible for managing and reporting the progress of a hierarchical
+     * delta-debugging process on a file tree structure. It utilizes a sequential progress reporter to
+     * keep track of the current progress and update it during the process. This class computes levels
+     * of directories and files to provide meaningful progress reporting.
+     *
+     * @property reporter An instance of [SequentialProgressReporter] used for updating the progress.
+     * @property currentLevel A volatile counter indicating the current level of progress.
+     * @property levelMaxDepths A hash map that stores the maximum depth levels of paths relative to the root.
+     * @constructor Creates a new instance of [ProgressReporter] based on the given root path and the array of root files.
+     */
+    private class ProgressReporter(val reporter: SequentialProgressReporter, root: Path, roots: Array<VirtualFile>) {
+        @Volatile
+        private var currentLevel = 0
+        private val levelMaxDepths = HashMap<Path, Int>()
+
+        init {
+            computeLevels(root, roots)
+        }
+
+        private data class StackEntry(val file: VirtualFile, val level: Int, var next: Int = 0)
+
+        private fun computeLevels(root: Path, roots: Array<VirtualFile>) {
+            val stack = mutableListOf<StackEntry>()
+            roots.forEach {
+                stack.add(StackEntry(it, 1))
+            }
+
+            while (stack.isNotEmpty()) {
+                val entry = stack.last()
+
+                val children = entry.file.children
+                if (children.isNotEmpty()) {
+                    if (entry.next < children.size) {
+                        stack.add(StackEntry(children[entry.next], entry.level + 1))
+                        entry.next += 1
+                    } else {
+                        stack.removeLast()
+                        levelMaxDepths[entry.file.toNioPath().relativeTo(root)] =
+                            children.maxOf { levelMaxDepths[it.toNioPath().relativeTo(root)]!! }
+                    }
+                } else {
+                    stack.removeLast()
+                    levelMaxDepths[entry.file.toNioPath().relativeTo(root)] = entry.level
+                }
```

Comment:
I kinda do not understand what is going here. What is `next`?

Reply:
That's the index of the next child to iterate. It's a recursion simulation. 

I will rename it and add some comments